src/components/field/svg/FieldOverlayRoot.tsx

```diff
@@ -99,6 +109,48 @@ class FieldOverlayRoot extends Component<Props, State> {
     this.handleResize();
 
     this.fieldSelection().call(this.zoomBehavior).on("dblclick.zoom", null);
+
+    this.fieldSelection().on("contextmenu", (e) => {
+      this.context.model.uiState.setContextMenuMouseSelection(e);
+    });
+  }
+
+  private handleCloseContextMenu() {
+    this.context.model.uiState.setContextMenuMouseSelection(undefined);
+    this.context.model.uiState.setContextMenuSelectedWaypoint(undefined);
+    this.context.model.uiState.setContextMenuWaypointType(undefined);
+  }
+
+  private handleContextMenuSelection(contextMenuWaypointType: number) {
+    // User selects field without selecting a waypoint
+    if (this.context.model.uiState.contextMenuMouseSelection === undefined) {
+      return;
+    }
+
+    if (this.context.model.uiState.contextMenuSelectedWaypoint === undefined) {
+      this.createWaypoint(
+        new MouseEvent("contextmenu", {
+          clientX: this.context.model.uiState.contextMenuMouseSelection[0],
+          clientY: this.context.model.uiState.contextMenuMouseSelection[1]
+        }),
+        contextMenuWaypointType
+      );
+      this.context.model.uiState.setContextMenuMouseSelection(undefined);
+    }
+
+    // User selects a waypoint
+    else if (
+      this.context.model.uiState.contextMenuSelectedWaypoint !== undefined
```

Comment:
Do we need this to be an else if?

Reply:
Why would it not? There could be future conditions we could add, but I don't want to fully short circuit that yet.

Reply:
This check is redundant and just adds noise.

Reply:
right sorry, was looking at the code from my phone.